File: Lession 3/queue.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
int main()
{
    int n, m;
    cin >> n >> m;
    vector<vector<int>> graph(n);
    vector<bool> visited(n, false);
    for (int i = 0; i < m; i++)
    {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    int count = 0;
    for (int i = 0; i < n; i++)
    {
        if (!visited[i])
        {
            queue<int> q;
            q.push(i);
            visited[i] = true;
            while (!q.empty())
            {
                int u = q.front();
                q.pop();
                for (int j = 0; j < graph[u].size(); j++)
                {
                    int v = graph[u][j];
                    if (!visited[v])
                    {
                        q.push(v);
                        visited[v] = true;
                    }
                }
            }
            count++;
        }
    }
    cout << count << endl;
    return 0;
}

File: Lession 3/queue.py
from collections import deque

class Dequeue:
    def __init__(self):
        self.elements = deque()
    def add_front(self, element):
        self.elements.appendleft(element)
        print(f"Added {element} to the front")
    def enqueue(self, element):
        self.elements.append(element)
        print(f"Enqueued {element}")
    def dequeue(self):
        if not self.is_empty():
            item = self.elements.popleft()
            print(f"Dequeued {item}")
            return item
        else:
            print("Queue is empty")
            return None
    def peek(self):
        if not self.is_empty():
            return self.elements[0]
        else:
            print("Queue is empty")
            return None
    def is_empty(self):
        return len(self.elements) == 0
    def display(self):
        print(self.elements)
if __name__ == "__main__":
    queue = Dequeue()
    queue.enqueue(1)
    queue.enqueue(2)
    queue.enqueue(3)
    queue.display()
    queue.dequeue()
    queue.display()
    queue.dequeue()
    queue.display()
    queue.dequeue()
    queue.display()
    queue.dequeue()
    queue.display()
    queue.enqueue(4)
    queue.display()
    queue.enqueue(5)
    queue.display()
    queue.enqueue(6)
    queue.display()
    queue.dequeue()
    queue.display()
    queue.dequeue()
    queue.display()
    queue.dequeue()
    queue.display()
    queue.dequeue()
    queue.display()
    queue.dequeue()
    queue.display()
